crypto/bls_core.c

```diff
@@ -341,51 +372,68 @@ typedef struct st_node {
 
 static node* new_node(const ep2_st* pk, const ep_st* sig){
     node* t = (node*) malloc(sizeof(node));
-    t->pk = (ep2_st*)pk;
-    t->sig = (ep_st*)sig;
-    t->right = t->left = NULL;
+    if (t) {
+        t->pk = (ep2_st*)pk;
+        t->sig = (ep_st*)sig;
+        t->right = t->left = NULL;
+    }
     return t;
 }
 
+static void free_tree(node* root) {
+    if (!root) return;
+
+    // only free pks and sigs of non-leafs, data of leafs are allocated 
+    // as an entire array in `bls_batchVerify`.
+    if (root->left) {   // no need to check the right child for the leaf check because
+                        //  the recursive build starts with the left side first
+        // relic free 
+        if (root->sig) ep_free(root->sig);
+        if (root->pk) ep2_free(root->pk);
+        // pointer free
+        free(root->sig);
+        free(root->pk);
+        // free the children nodes
+        free_tree(root->left);
+        free_tree(root->right);
+    }
+    free(root);
+}
+
 // builds a binary tree of aggregation of signatures and public keys recursively.
 static node* build_tree(const int len, const ep2_st* pks, const ep_st* sigs) {
     // check if a leaf is reached
     if (len == 1) {
-        return new_node(pks, sigs);  // use the first element of the arrays
+        return new_node(&pks[0], &sigs[0]);  // use the first element of the arrays
     }
 
     // a leaf is not reached yet, 
     int right_len = len/2;
     int left_len = len - right_len;
 
     // create a new node with new points
-    node* t = new_node((ep2_st*)malloc(sizeof(ep2_st)), (ep_st*)malloc(sizeof(ep_st)));
+    ep2_st* new_pk = (ep2_st*)malloc(sizeof(ep2_st));
+    if (!new_pk) return NULL;
+    ep_st* new_sig = (ep_st*)malloc(sizeof(ep_st));
+    if (!new_sig) { free(new_pk); return NULL; }
+
+    node* t = new_node(new_pk, new_sig);
+    if (!t) { free(new_sig); free(new_pk); return NULL; }
     ep_new(t->sig);
-    ep_new(t->pk);
+    ep2_new(t->pk);
+
     // build the tree in a top-down way
     t->left = build_tree(left_len, &pks[0], &sigs[0]);
+    if (!t->left) { free_tree(t); return NULL; }
+
     t->right = build_tree(right_len, &pks[left_len], &sigs[left_len]);
+    if (!t->right) { free_tree(t); return NULL; }
     // sum the children
     ep_add_jacob(t->sig, t->left->sig, t->right->sig);
     ep2_add_projc(t->pk, t->left->pk, t->right->pk); 
     return t;
```

Comment:
You may consider adding a separate error handling trailer after the `return t;`:
```
err:
    free(new_sig);
errSig:
    free(new_pk);
errPk:
    return NULL;
```

that way we'll be slightly more consistent.